hollow/src/main/java/com/netflix/hollow/api/producer/HollowIncrementalCyclePopulator.java

```diff
@@ -138,15 +144,32 @@ private void removeRecordsFromNewState(HollowProducer.WriteState newState, Map<S
     }
 
     private void addRecords(final HollowProducer.WriteState newState) {
+        List<Map.Entry<RecordPrimaryKey, Object>> entryList = new ArrayList<>(mutations.entrySet());
+        
+        AtomicInteger nextMutation = new AtomicInteger(0);
+        
         SimultaneousExecutor executor = new SimultaneousExecutor(threadsPerCpu);
-        for(final Map.Entry<RecordPrimaryKey, Object> entry : mutations.entrySet()) {
-            executor.execute(new Runnable() {
-                public void run() {
-                    if(entry.getValue() != DELETE_RECORD)
-                        newState.add(entry.getValue());
+        executor.execute(() -> {
+            FlatRecordDumper flatRecordDumper = null;
```

Comment:
There is no parallelism since only one task is executed asynchronously and the code that follows waits for that task in the executor to complete. Was that the intention?
My suggestion would be to keep this simple, snapshot the mutations (`values()`) into a list, and operate sequentially within the same thread. We can easily change this later to a stream and operate in parallel getting good decomposition. Or alternatively split the list into N parts in proportion to the number of hardware cores you want to use and execute N times (but parallel streams can do this more easily and better).
The guide of "threadsPerCpu" to control the parallelism is likely to be wrong most of the time. We need to express something that takes into account the machine parallelism (number of hardware cores to use), the amount of work (number of mutations), and the unit cost per work (the cost for each mutation). This is not easy to get right, hence why there is currently no mechanism in parallel streams :-) however we can control the former (machine parallelism) by executing the parallel stream within a custom fork join pool (to avoid using the common fork join pool).